File: EjemplosConcurrencia/CSharp/EjemploSemaforos/EjemploSemaforos/EjemploSemaforos/Productor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace EjemploSemaforos
{
    public class Productor
    {
        private ListBox modeloProductor;
        private delegate void AgregarItem();
        public Productor(ListBox modeloProductor)
        {
            this.modeloProductor = modeloProductor;
        }
        public void run()
        {
            try
            {
                while (true)
                {
                    frmEjemploSemaforos.semaforo2.WaitOne();
                    producir();
                    frmEjemploSemaforos.semaforo1.Release();
                    Thread.Sleep(1000);
                }
            }catch(Exception ex)
            {
                MessageBox.Show("Ha ocurrido una excepción: " + ex.Message);
            }
        }

        public void producir()
        {
            Elemento e = new Elemento();
            modeloProductor.Invoke(new AgregarItem(()=> modeloProductor.Items.Add(e)));
        }
    }
}

File: EjemplosConcurrencia/JAVA/EjemplosConcurrencia/src/main/java/pe/edu/pucp/softprog/semaforos/Productor.java
package pe.edu.pucp.softprog.semaforos;
import javax.swing.DefaultListModel;
public class Productor extends Thread{
    
    private DefaultListModel modelo;
    
    public Productor(DefaultListModel modelo){
        this.modelo = modelo;
    }
    
    @Override
    public void run(){
        try{
            while(true){
                frmEjemplo.semaforo2.acquire();
                producir();
                frmEjemplo.semaforo1.release();
                Thread.sleep(1000);
            }
        }catch(Exception ex){
            System.out.println("ERROR AL PRODUCIR: " +
                    ex.getMessage());
        }
    }
    
    public void producir() throws Exception{
        Elemento elemento = new Elemento();
        modelo.addElement(elemento);
    }
}
